### DISCUSSION TITLE
MCR-ALS with kinetic constraints with "A -> B" reaction

### DISCUSSION THREAD

**Discussion by liuchzzyy** [spectrochempy/spectrochempy#722] [Q&A] [answered]
Dear all,
I am interested in the module of MCR-ALS with kinetic constraints, but being troublesome in this type of "A -> B" reaction.
the related codes like below:

''' kinetic model '''
reactions = ("A -> B")
species_concentrations = {"A": 0.69, "B": 0.31}
k0 = np.array((0.5, 0.05))
kin = scp.ActionMassKinetics(reactions, species_concentrations, k0)

However, it shows: ERROR | ValueError: arrhenius should contain 0 rate constants, 2 have been provided 

and I tried others like below:
k0 = np.array((0.5))
it shows:  ERROR | IndexError: tuple index out of range

k0 = np.empty((0))
it shows:  ERROR | TypeError: iteration over a 0-d array

I am wondering if I need to set 1 rate constants for this equation , but why deos 0 rate constants show here? and or...
by the way, the example in spectrochempy.fr, works well

Thank you!
Cheng

**Comment by atravert** [comment, marked as answer]
Hi Cheng,

here  are the reasons of the error:
1/ `reactions` should be  a `list`, a `tuple` or a `dict`, but `("A -> B")` is a string so, the MassActionKinetics instance contains 0 reaction and so expects 0 rate constant
2/  `arrhenius` (your `k0` variable) is a 1D array of size 2, hence you provide 2 rate constants.

The following script should work:

```
import numpy as np
import spectrochempy as scp

# kinetic model 
reactions = ("A -> B",)  # note the coma: this is now a tuple of size 1    
species_concentrations = {"A": 0.69, "B": 0.31}
k0 = np.array([0.5])  # as we have a single reaction, we must use a single rate constant. 
kin = scp.ActionMassKinetics(reactions, species_concentrations, k0)

# integrate and plot concentratuon profiles
t = np.linspace(0, 10, 100)
C = kin.integrate(t)
C.T.plot()
scp.show()   
```

And here is the output:

![image](https://github.com/spectrochempy/spectrochempy/assets/10098314/00eaef0a-f903-49b7-9593-ca1070b9dc8c)


I hope this helps.
Arnaud

**Reply by atravert** [reply]
PS: in the next release we will add a type and size validation for `reactions`, so that a more explicit error message will be given. Thank you for your contribution !
A.

**Reply by liuchzzyy** [reply]
Thanks again! and it works now. the comma does matter!
